File: src/main/java/com/polarbookshop/orderservice/order/event/OrderFunctions.java
package com.polarbookshop.orderservice.order.event;

import com.polarbookshop.orderservice.order.domain.OrderService;
import java.util.function.Consumer;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import reactor.core.publisher.Flux;

@Configuration
@Slf4j
public class OrderFunctions {

  @Bean
  public Consumer<Flux<OrderDispatchedMessage>> dispatchOrder(OrderService orderService) {
    return flux ->
        orderService
            .consumeOrderDispatchedEvent(flux)
            .doOnNext(order -> log.info("Order with id {} is dispatched", order.id()))
            .subscribe();
  }
}

File: src/main/kotlin/com/polarbookshop/orderservice/order/event/OrderFunctions.kt
package com.polarbookshop.orderservice.order.event

import com.polarbookshop.orderservice.order.domain.OrderService
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import reactor.core.publisher.Flux
import java.util.function.Consumer

private val logger = KotlinLogging.logger {}

@Configuration
class OrderFunctions {
    @Bean
    fun dispatchOrder(orderService: OrderService): Consumer<Flux<OrderDispatchedMessage>> =
        Consumer { messages ->
            orderService
                .consumeOrderDispatchedEvent(messages)
                .doOnNext { order ->
                    logger.info { "The order with id $order is dispatched" }
                }.subscribe()
        }
}
